File: python/longest_common_prefix.py
import argparse
from typing import List


class Solution:
    def __init__(self):
        self.maxprefix = None
        self.longest_prefix = None

    def longest_common_prefix(self, strs: List[str]) -> str:
        self.maxprefix = strs[0]
        if not self.maxprefix:
            return ""
        if len(strs) == 1:
            return self.maxprefix
        for word in strs[1:]:
            current_length = 0
            iterate_to = min(len(word), len(self.maxprefix))
            for i in range(iterate_to):
                if self.maxprefix[i] == word[i]:
                    current_length += 1
                else:
                    self.set_current_longest_prefix(current_length)
                    break
            self.set_current_longest_prefix(current_length)
        return self.maxprefix[:self.longest_prefix]

    def set_current_longest_prefix(self, current_length: int) -> None:
        if self.longest_prefix is None:
            self.longest_prefix = current_length
        else:
            self.longest_prefix = min(self.longest_prefix, current_length)


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('--input', nargs="+", type=str)
    args = parser.parse_args()
    return args.input


def main():
    strs = parse_args()
    result = Solution().longest_common_prefix(strs)
    print(result)


if __name__ == "__main__":
    main()

File: go/longestCommonPrefix.go
package main

import (
	"flag"
	"fmt"
)

func setLength(lenght int, curr int) int {
	if lenght == -1 {
		return curr
	} else {
		return min(lenght, curr)
	}
}

func longestCommonPrefix(strs []string) string {
	base := strs[0]
	if base == "" {
		return ""
	}
	if len(strs) == 1 {
		return base
	}
	lenght := -1
	for _, word := range strs[1:] {
		currLenght := 0
		toIterate := min(len(base), len(word))
		for i := 0; i < toIterate; i++ {
			if string(word[i]) == string(base[i]) {
				currLenght++
			} else {
				lenght = setLength(lenght, currLenght)
				break
			}
		}
		lenght = setLength(lenght, currLenght)
	}
	return base[:lenght]
}



func main() {
	funcToRun := flag.String("problem", "", "Defines the problem to solve")
	flag.Parse()
	fmt.Println(*funcToRun)
	strs := flag.Args()
	result := longestCommonPrefix(strs)
	fmt.Printf("Result is: %s\n", result)
}
